src/action/install_date.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright 2020 Google LLC
+//
+// Use of this source code is governed by an MIT-style license that can be found
+// in the LICENSE file or at https://opensource.org/licenses/MIT.
+
+//! A handler and associated types for the install date action.
+//!
+//! The install date action uses various heuristics to detect the operating
+//! system install date.
+
+use std::time::{SystemTime, Duration};
+
+use log::error;
+use crate::session::{self, Session};
+
+#[cfg(target_family = "unix")]
+use std::{fs, path::Path};
+
+/// A response type for the install date action.
+struct Response {
+    /// Install date of the operating system, or `None` if the attemps to
+    /// obtain install date failed.
+    time: Option<SystemTime>,
+}
+
+impl super::Response for Response {
+
+    // TODO: change to RDFDatetime when the client will be capable to send
+    // "raw" strings without wrapping them into protobuf
+    const RDF_NAME: Option<&'static str> = Some("DataBlob");
+
+    type Proto = u64;
+
+    fn into_proto(self) -> Self::Proto {
+        let time = match self.time {
+            Some(time) => time,
+            None => {
+                error!("cannot get install date, all methods failed");
+                std::time::UNIX_EPOCH
+            },
+        };
+        let since_epoch = match time.duration_since(std::time::UNIX_EPOCH) {
+            Ok(duration) => duration,
+            Err(err) => {
+                error!(
+                    "install date is {} seconds earlier than Unix epoch",
+                    err.duration().as_secs()
+                );
+                Duration::from_secs(0)
+            },
+        };
+        since_epoch.as_secs()
+    }
+}
+
+/// Iterates over files given it `iter`, checking their modification times.
+///
+/// If the function cannot get modification time from all the files (e. g.
+/// they all don't exist), the return value is `None`. Otherwise, the
+/// modification time of the first file is returned.
+#[cfg(target_family = "unix")]
+fn get_modified_time<P, It>(iter: It) -> Option<SystemTime>
+where
+    P: AsRef<Path>,
+    It: Iterator<Item = P>,
+{
+    for path in iter {
+        let time = fs::metadata(path).and_then(|metadata| metadata.modified());
+        if let Ok(time) = time {
+            return Some(time);
+        }
+    }
+
+    None
```

Comment:
I think this could be simplified to this:

```rust
iter.filter_map(|path| fs::metadata(path).modified().ok()).next()
```

Although I am not sure if it is more readable ¯\\\_(ツ)\_/¯